# D1AED: Análise Estatística para Ciência de Dados (IFSP Campinas)

## Atividade em Grupo 1 - Prazo: 09/05/2023 às 23:59
Nesta atividade, cada grupo (**máximo de 3 integrantes**) deverá resolver a lista de exercícios presentes neste notebook. Todos os exercícios deverão ser respondidos no própio notebook usando Python. 

#### Submissão
- Cada grupo deverá submeter um **arquivo .zip** contendo o _notebook_ com as respostas e a pasta `datasets`, contendo todos os datasets usados no notebook;
- A submissão da atividade será feita em tarefa específica no Moodle da disciplina;
- Apenas **um membro do grupo** deverá submeter a atividade;
- Há materiais de apoio/estudos na tarefa do Moodle.

**Prazo:** 09/05/2023 às 23:59

#### **Nomes e Prontuários dos Integrantes do Grupo (máx 3):**
- João Pedro de Oliveira Ferreira (CP3021696)


In [ ]:
import random
import numpy as np

random.seed(42)
np.random.seed(seed=42)

### Questão 1
At a given university, forty percent of students speak three languages. When randomly selecting 15 students:

**(a)** What is the probability that 3 students speak three languages?

In [9]:
#Importando da biblioteca scipy.stats as funções de distribuição binomial

from scipy.stats import binom

#Definindo a probabilidade de sucesso e fracasso

#sucesso

p = 0.4

#fracasso

q = 1 - p

#Definindo os parâmetros que iremos utilizar para calcular a PMF

n = 15
k = 3

#Calculando a PMF

prob = binom.pmf(k, n, p)

print(f'''A probabilidade de escolher 15 estudantes e 3 falarem três línguas é de {round(prob*100,2)}%''')

A probabilidade de escolher 15 estudantes e 3 falarem três línguas é de 6.34%


**(b)** What is the probability that 12 students do not speak three languages?

In [10]:
# Podemos considerar que a nossa probabilidade de sucesso agora é a probabilidade de fracasso no exercício anterior

#Calculando diretamente a PMF

new_prob = binom.pmf(k = 12, n = 15, p = 0.6)

print(f'''A probabilidade de escolher 15 estudantes e 12 não falarem três línguas é de {round(new_prob*100,2)}%''')

A probabilidade de escolher 15 estudantes e 12 não falarem três línguas é de 6.34%


**(c)** What is the probability that at least 3 students speak three languages?

In [29]:
# Podemos calcular a probabilidade de 3,4,5,6,7,8,9,10,11,12,13,14 e 15


prob_3 = binom.pmf(k = 3, n = 15, p = 0.4)
prob_4 = binom.pmf(k = 4, n = 15, p = 0.4)
prob_5 = binom.pmf(k = 5, n = 15, p = 0.4)
prob_6 = binom.pmf(k = 6, n = 15, p = 0.4)
prob_7 = binom.pmf(k = 7, n = 15, p = 0.4)
prob_8 = binom.pmf(k = 8, n = 15, p = 0.4)
prob_9 = binom.pmf(k = 9, n = 15, p = 0.4)
prob_10 = binom.pmf(k = 10, n = 15, p = 0.4)
prob_11 = binom.pmf(k = 11, n = 15, p = 0.4)
prob_12 = binom.pmf(k = 12, n = 15, p = 0.4)
prob_13 = binom.pmf(k = 13, n = 15, p = 0.4)
prob_14 = binom.pmf(k = 14, n = 15, p = 0.4)
prob_15 = binom.pmf(k = 15, n = 15, p = 0.4)

prob_final = prob_3+prob_4+prob_5+prob_6+prob_7+prob_8+prob_9+prob_10+prob_11+prob_12+prob_13+prob_14+prob_15

print("Via probabilidade >= 3:")
print(f'''A probabilidade de escolher 15 estudantes e pelo menos 3 falarem três línguas é de {round(prob_final*100,2)}%\n''')

# Ou podemos calcular a probabilidade de 0, 1, 2 e subtrair de 100%

prob_0 = binom.pmf(k = 0, n = 15, p = 0.4)
prob_1 = binom.pmf(k = 1, n = 15, p = 0.4)
prob_2 = binom.pmf(k = 2, n = 15, p = 0.4)

prob_final2 = prob_0+prob_1+prob_2

print("Via probabilidade < 3:")
print(f'''A probabilidade de escolher 15 estudantes e pelo menos 3 falarem três línguas é de {round(100-prob_final2*100,2)}%\n''')

# Ou poderíamos calcular a CDF

prob_cdf = 1 - binom.cdf(k = 2, n = 15, p = 0.4)

print("Via CDF:")
print(f'''A probabilidade de escolher 15 estudantes e pelo menos 3 falarem três línguas é de {round(prob_cdf*100,2)}%''')


Via probabilidade >= 3:
A probabilidade de escolher 15 estudantes e pelo menos 3 falarem três línguas é de 97.29%

Via probabilidade < 3:
A probabilidade de escolher 15 estudantes e pelo menos 3 falarem três línguas é de 97.29%

Via CDF:
A probabilidade de escolher 15 estudantes e pelo menos 3 falarem três línguas é de 97.29%


### Questão 2

The mean of defects in ethernet cables manufactured by a company is 1.2 defects for every 6 meters of length. 
What is the probability of getting less than 2 defects when inspecting 6 meters of cable at random?

In [38]:
#Neste exercício iremos utilizar a distribuição de Poisson

from scipy.stats import poisson

mu = 1.2/6

#Utilizando PMF temos

prob_0 = poisson.pmf(k = 0, mu = 1.2/6)
prob_1 = poisson.pmf(k = 1, mu = 1.2/6)

prob_final = 1 - prob_0 - prob_1

print(f'''A probabilidade de ao inspecionar 6 metros de cabo e ter menos que 2 defeitos é de {round(prob_final*100,2)}%''')

A probabilidade de ao inspecionar 6 metros de cabo e ter menos que 2 defeitos é de 1.75%


### Questão 3

If we consider the distribution of blood types in Hawkins —— where 30% of the population has O+ blood type, 33% has A+, 12% has B+, 6% has AB+, 7% has O-, 8% has A-, 3% has B-, and 1% has AB- —— what is the probability that a group of 15 randomly chosen Austrian citizens will consist of 3 individuals with O+ blood type, 2 with A+, 3 with B+, 2 with AB+, 1 with O- blood type?

In [51]:
#Importando a biblioteca de probabilidade multinomial

from scipy.stats import multinomial

#Declarando as probabilidades individualmente

prob_o_positive = 0.30
prob_a_positive = 0.33
prob_b_positive = 0.12
prob_ab_positive = 0.06
prob_o_negative = 0.07
prob_a_negative = 0.08
prob_b_negative = 0.03
prob_ab_negative = 0.01

#Declarando o vetor de probabilidade

vetor_prob = [prob_o_positive,prob_a_positive,prob_b_positive,prob_ab_positive,prob_o_negative,prob_a_negative,prob_b_negative,prob_ab_negative]

#Declarando o vetor de quantidade 

qtd = [3, 2, 3, 2, 1, 0, 0, 0]

#Perceba que o número total do vetor qtd é de 11 e não 15 conforme o enunciado. Podemos fazer um truque:

new_qtd = [3, 2, 3, 2, 1, 4]

vetor_prob_new = [prob_o_positive,prob_a_positive,prob_b_positive,prob_ab_positive,prob_o_negative,prob_a_negative+prob_b_negative+prob_ab_negative]

#Calculando a probabilidade

prob = multinomial.pmf(x = new_qtd, n = 15, p = vetor_prob_new)

print(f'''A probabilidade será de {round(prob*100,5)}%''')

A probabilidade será de 0.01005%


### Questão 4

Patients with a serious illness have been treated by a drug that causes a side effect at a rate of three patiences in a hundred. When selecting ten patients treated with this drug, what is the chance that at least one will have side effects?

In [56]:
#Probabilidade de efeito colateral 

p = 3/100
q = 97/100

#Calculando as probabilidades para 0 e tirando de 100%

prob = binom.pmf(k = 0 , n = 10, p = p)

print(f'''A probabilidade será de {round((1-prob)*100,2)}%''')

A probabilidade será de 26.26%


### Questão 5

The length of a human pregnancy approximates a normal distribution with a mean of 272 days with a standard deviation of 9 days.

**(a)** What proportion of all pregnancies will last between 265 and 280 days?

In [67]:
from scipy.stats import norm

#De acordo com a documentação o parâmetro loc é a média e scale é o devio padrão

prob_280 = norm.cdf(x= 280, loc = 272, scale = 9)
prob_265 = norm.cdf(x = 265, loc = 272, scale = 9)

print(f'''A probabilidade será de {round(100*(prob_280-prob_265),2)}%''')

A probabilidade será de 59.46%


**(b)** What length of time marks the shortest 70% of all pregnancies? 

In [72]:
# De acordo com a tabela Z o valor para uma área de 0.7 é de aproximadamente 0.53. Logo

x = 0.53*9+272

print(f'''{x} dias''')

276.77 dias


### Questão 6

The average numbers of traffic accidents on a certain section of highway is 0.5 per week. <br/>
Find the probability that:

**(a)** There will be less than 2 accidents in a week selected at random:

In [74]:
from scipy.stats import poisson

prob_0 = poisson.pmf(k = 0, mu = 0.5)
prob_1 = poisson.pmf(k = 1, mu = 0.5)

prob_final = prob_0 + prob_1

print(f'''A probabilidade será de {round(prob_final*100,2)}%''')

A probabilidade será de 90.98%


**(b)** In a particular week, there will be more than 2 accidents:

In [75]:
prob_0 = poisson.pmf(k = 0, mu = 0.5)
prob_1 = poisson.pmf(k = 1, mu = 0.5)

prob_final = 1 - prob_0 - prob_1

print(f'''A probabilidade será de {round(prob_final*100,2)}%''')

A probabilidade será de 9.02%


**(c)** There will be no accidents in three weeks:

In [76]:
# 0.5 acidentes por semana, logo teremos 1.5 acidentes por 3 semanas

prob_0 = poisson.pmf(k = 0, mu = 1.5)

prob_final = prob_0

print(f'''A probabilidade será de {round(prob_final*100,2)}%''')

A probabilidade será de 22.31%


### Questão 7

Suppose you are drawing cards randomly from a standard deck with 52 cards, replacing each card after drawing, and repeating the process five times. What is the likelihood of drawing one heart, one spade, one club, and two diamonds?

### Questão 8

A factory produces chocolates in boxes of 200. The probability of missing one chocolate in a box is 0.006. What is the probability that a randomly selected box is missing at most two chocolates? <br/>
**Tip:** [Poisson Approximation to the Binomial.](https://bestmaths.net/online/index.php/year-levels/year-12/year-12-topic-list/poisson-approximation-binomial/#:~:text=Poisson%20Approximation%20to%20the%20Binomial&text=When%20the%20value%20of%20n,Poisson%20is%20a%20good%20approximation.)

### Questão 9
In a town, the distance traveled by students to school is normally distributed with $\mu=7.5$km and $\sigma=2.5$km.

**(a)** What is the probability that the distance covered by a randomly selected student is less than 11 km?

**(b)** What is the probability that the distance covered by a randomly selected student is betweem 5.5 km than 10.5 km?

**(c)** Find _d_ such that 10% of the students travel less than _d_ kilometers to school.

**(d)** What is the probability that the mean distance traveled by a random sample of 6 students is less than 5 km? <br/>
**Tip:** Central Limit Theorem.

### Questão 10

Suppose a recent poll showed that among the respondents, 23% supported candidate A, 19% supported candidate B, 13% supported candidate C, and the rest were undecided. If seven individuals are randomly selected, what is the likelihood that 2 of them support candidate A, 2 support candidate B, and 1 supports candidate C?

### Questão 11

The recovery time of a patient with a certain illness follows a normal distribution with a standard deviation of 2 days. Calculate the confidence level for a sample of 12 patients which indicates that the population recovery time is between 13 and 15 days.

### Questão 12

John moved out to _Stats City_ for work. John needs to rent a place to live, so he has compiled a list of 60 apartments randomly chosen from the EasyRent real estate. This compilation is stored at the dataset `'./datasets/rents.csv'`, where the column `price` is the monthly rent price in dollars.

**(a)** Construct a 95% confidence interval by Central Limit Theorem and Bootstrapping. Compare the results.

**(b)** To what population of apartments can you appropriately infer from your sample?

We can most accurately infer to Stats Cits apartments listed on EasyRent. <br/>
We should not infer to _all_ apartments in Stats City because we do not know if apartments on EasyRent are representative of all apartments in Stats City.

**(c)** How large a sample of apartments be needed to estimate the population mean within $\pm$ \$50.0 with 90% confidence?

### Questão 13

In order to estimate the average height of the population of a town, a survey collected a sample containing the height of 400 people chosen at random. The file `'./datasets/heights.csv'` contains the registered heights.

The average heights of a random sample of 400 people from a city is 1.75 m. It is known that the heights of the population are random variables that follow a normal distribution with a variance of 0.16.

**(a)** Construct a 95% confidence interval by using the Central Limit Theorem and Bootstrapping. Compare the results.

**(b)** With a confidence level of 90%, what would the minimum sample size need to be in order for the true mean of the heights to be less than 2 cm from the sample mean?

### Questão 14

Monthly sales for a bakery follow a normal distribution with a standard deviation of \\$900.00. By statistically studying sales in the last 9 months, a confidence interval for the mean monthly sales of [\\$4,663.00, \\$5,839.00] was found.

**(a)** What were the average sales over the nine month period?

**(b)** What is the confidence level for this interval?

### Questão 15

A survey was carried out to find out if Brazilians believe that Brazil will win the next world cup. A sample of 1000 Brazilians spread across the country and chosen at random was observed. Your result has been stored in file `./datasets/world_cup.csv`.

Construct a 95% confidence interval on the Brazilian population who believes Brazil will win the next world cup. Use the method based on the Central Limit Theorem, Bootstrapping, and compare the results.

**Tips**:
- [Confidence Intervals for a Population Proportion](https://online.stat.psu.edu/stat100/lesson/9/9.1)
- [statsmodels.stats.proportion.proportion_confint](https://www.statsmodels.org/dev/generated/statsmodels.stats.proportion.proportion_confint.html)

### Questão 16

The municipal administration wants to estimate the proportion of color blind in the city. For that, they observed a random sample of individuals, where the data is stored in the dataset `'./datasets/color_blind.csv'`. The column `color_blind` tells whether or not the person is colorblind.

**(a)** Construct a confidence interval for the proportion of the color blind population using a confidence level of 99%. Use the method based on the Central Limit Theorem, Bootstrapping, and compare the results.

**(b)** Consider that the the percentage of color blind people in the sample is 30%. Estimate the value of _n_ so that, with a significance level of 5%, the error in the estimate is less than 3.1%.

### Questão 17

The GoodFood restaurant offers _lunch_ and _dinner_ meals. There are two types of drinks: _juice_ and _soda_. GoodFood's manager suspects customers drink more soda at dinner. The manager randomly selects data from the system and the data was stored in the dataset `./datasets/goodfood.csv`. Each row contains the information of the meal time (column `meal`) and type of drink (column `drink`) ordered by the customer.

Construct a 98% confidence interval for the proportion of ordering _soda_ for _dinner_ hours. Use the method based on the Central Limit Theorem, Bootstrapping, and compare the results.